src/main/java/org/spongepowered/common/mixin/core/world/MixinChunk.java

```diff
@@ -886,4 +933,88 @@ public int getBlockDigTimeWith(int x, int y, int z, ItemStack itemStack, Cause c
         return this.world.getBlockDigTimeWith(this.xPosition << 4 + (x & 15), y, this.zPosition << 4 + (z & 15), itemStack, cause);
     }
 
+    @Override
+    public void setNeighbor(Direction direction, @Nullable Chunk neighbor) {
+        this.neighbors[directionToIndex(direction)] = neighbor;
+    }
+
+    @Override
+    public Optional<Chunk> getNeighbor(Direction direction, boolean shouldLoad) {
+        checkNotNull(direction, "direction");
+        checkArgument(!direction.isSecondaryOrdinal(), "Secondary cardinal directions can't be used here");
+
+        int index = directionToIndex(direction);
+        Direction secondary = getSecondaryDirection(direction);
+        Chunk neighbor = this.neighbors[index];
+
+        if (neighbor == null && shouldLoad) {
+            Vector3i neighborPosition = this.getPosition().add(getCardinalDirection(direction).toVector3d().toInt());
+            Optional<Chunk> cardinal = this.getWorld().loadChunk(neighborPosition, true);
+            if (cardinal.isPresent()) {
+                neighbor = cardinal.get();
+            }
+        }
+
+        if (neighbor != null) {
+            if (secondary != Direction.NONE) {
+                return neighbor.getNeighbor(secondary, shouldLoad);
+            } else {
+                return Optional.of(neighbor);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private static int directionToIndex(Direction direction) {
+        switch (direction) {
+            case NORTH:
+            case NORTHEAST:
+            case NORTHWEST:
+                return 0;
+            case SOUTH:
+            case SOUTHEAST:
+            case SOUTHWEST:
+                return 1;
+            case EAST:
+                return 2;
+            case WEST:
+                return 3;
+            default:
+                throw new IllegalArgumentException("Unexpected direction");
```

Comment:
Just thinking that `Direction.UP` and `DOWN` are not 'unexpected', but rather unsupported. Perhaps `getNeighbor` should gracefully return `Optional.empty()` if you attempt to get a chunk in the y-plane (since it's unsupported in vanilla)